packages/flutter/lib/src/material/input.dart

```diff
@@ -13,6 +13,126 @@ import 'theme.dart';
 
 export 'package:sky_services/editing/editing.mojom.dart' show KeyboardType;
 
+/// A container for grouping together multiple Input widgets.
+class Form extends StatefulWidget {
+  Form({
+    Key key,
+    this.child,
+    this.data,
+    this.onSubmitted
+  }) : super(key: key) {
+    assert(child != null);
+  }
+
+  /// A map of form field IDs to their input values.
+  final Map<String, String> data;
+
+  /// Called when the input is accepted anywhere on the form.
+  final VoidCallback onSubmitted;
+
+  /// Root of the widget hierarchy that contains this form.
+  final Widget child;
+
+  @override
+  _FormState createState() => new _FormState();
+}
+
+class _FormState extends State<Form> {
+  // A map of form field IDs to their state.
+  final Map<String, _FormFieldState> fieldState = <String, _FormFieldState>{};
+
+  void onChanged(String id, InputValue value) {
+    config.data[id] = value.text;
+    // We sometimes need other inputs to respond to changes (e.g. a password
+    // verifier should update its error text if the other password field
+    // changes).
+    fieldState.forEach((String key, _FormFieldState data) {
+      data.inputState.setState((){ });
+    });
+  }
+  void onSubmitted(String id, InputValue value) {
+    config.onSubmitted();
+  }
+
+  Widget build(BuildContext context) {
+    return new _FormScope(
+      state: this,
+      child: config.child
+    );
+  }
+}
+
+typedef String FormFieldValidator(InputValue);
+
+/// This contains identifying information for Input fields, required if the
+/// Input is part of a Form.
+class FormField {
+  FormField({
+    this.id,
+    this.validator
+  }) {
+    assert(id != null);
+  }
+
+  /// An identifier for this Input widget. Should be unique among Inputs in this
+  /// Form.
+  final String id;
+
+  /// An optional method that validates an input. Returns an error string to
+  /// display if the input is invalid, or null otherwise.
+  final FormFieldValidator validator;
+}
+
+class _FormScope extends InheritedWidget {
+  _FormScope({
+    Key key,
+    Widget child,
+    this.state
+  }) : super(key: key, child: child);
+
+  _FormState state;
+  Map<String, _FormFieldState> get fieldState => state.fieldState;
+
+  // TODO(mpcomplete): not sure what to compare here.
+  @override
+  bool updateShouldNotify(_FormScope old) {
+    return fieldState != old.fieldState;
+  }
+}
+
+class _FormFieldState {
+  _FormFieldState(this.scope, this.id, this.inputState);
+
+  final _FormScope scope;
+  final String id;
+  final _InputState inputState;
+  InputValue value = new InputValue();
+
+  static _FormFieldState at(BuildContext context, FormField field, _InputState inputState) {
+    _FormScope result = context.inheritFromWidgetOfExactType(_FormScope);
+    if (result != null) {
+      assert(field != null);
+      result.fieldState.putIfAbsent(field.id,
```

Comment:
* by "move around" i mean consider if in a subsequent build, the Input isn't included.

* InheritedWidget's sole purpose is to make people who hook into it rebuild. See, for example, IconTheme. Whenever someone uses IconTheme.of(), they get added to a list such that the next time the IconTheme is build, all the people on the list are also rebuilt. (That's what "updateShouldNotify" is about. If it returns true, then all the clients are rebuilt.)

* I'm not sure what you mean by your # 1.